File: frontend/src/lib/messaging.ts
// Messaging channels API client — /api/messaging/* (per-tenant SMS/Telegram/WhatsApp).
// Mirrors lib/mail.ts / lib/billing.ts: bget for never-throw reads, bpost/bpatch/bdel for actions.
// Secrets are write-only — create/patch send a token only when newly typed; reads expose has_token.
import { bget, bpost, bpatch, bdel, type Fetched } from './billing'

export type ChannelKind = 'SMS' | 'TELEGRAM' | 'WHATSAPP'

export type ChannelAccount = {
  id: string
  channel: ChannelKind
  provider: string
  display_name: string
  sender_id?: string | null
  is_default: boolean
  is_active: boolean
  status: string
  last_error?: string | null
  config?: Record<string, any>
  has_token: boolean
  has_extra: boolean
}

export type ChannelAccountInput = {
  channel: ChannelKind
  display_name: string
  provider?: string
  sender_id?: string
  secret_token?: string          // write-only; sent only when newly typed
  secret_extra?: string
  config?: Record<string, any>
  is_default?: boolean
  is_active?: boolean
}

export type ChannelTestResult = { ok: boolean; status: string; detail?: string | null }

export function listAccounts(token: string): Promise<Fetched<ChannelAccount[]>> {
  return bget<ChannelAccount[]>(token, '/api/messaging/accounts')
}
export function createAccount(token: string, data: ChannelAccountInput): Promise<ChannelAccount> {
  return bpost<ChannelAccount>(token, '/api/messaging/accounts', data)
}
export function updateAccount(token: string, id: string, data: Partial<ChannelAccountInput>): Promise<ChannelAccount> {
  return bpatch<ChannelAccount>(token, `/api/messaging/accounts/${id}`, data)
}
export function deleteAccount(token: string, id: string): Promise<void> {
  return bdel(token, `/api/messaging/accounts/${id}`)
}
export function testAccount(token: string, id: string): Promise<ChannelTestResult> {
  return bpost<ChannelTestResult>(token, `/api/messaging/accounts/${id}/test`)
}
export function sendMessage(token: string, data: { channel: string; to: string; text: string; subject?: string }) {
  return bpost(token, '/api/messaging/send', data)
}

File: backend/app/models/messaging.py
"""Messaging channels — per-tenant SMS / Telegram / WhatsApp config (the Mail-module pattern).

Each tenant ISP configures its OWN channel credentials (Telegram bot token, Viva Armenia SMS creds,
WhatsApp Business token) — multi-tenant from day one: 5 tenants = 5 independent, RLS-isolated,
Fernet-encrypted rows. `channels.dispatch` routes a tenant's sms/telegram/whatsapp traffic through
ITS configured account (the `mail_account → SmtpEmailGateway` pattern applied to messaging).

Telegram ships live; Viva-SMS + WhatsApp gateways are registered stubs (creds/API pending from the
pilot) — the per-tenant framework is ready, they slot in when HouseNet shares access.

Enums (UPPER_SNAKE, B1):
  ChannelKind          : SMS | TELEGRAM | WHATSAPP
  ChannelAccountStatus : PENDING | CONNECTED | AUTH_ERROR | CONN_ERROR | DISABLED
"""
from app.utils.ids import uuid7
import uuid
from datetime import datetime

from sqlalchemy import String, Boolean, ForeignKey, DateTime, func, Text, Index, text
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base
from ..security import EncryptedString


class TenantChannelAccount(Base):
    """A tenant's outbound messaging account for one channel (SMS/Telegram/WhatsApp). Credentials are
    Fernet-encrypted at rest; the sender drives the channel from THIS row's creds, never a global env."""
    __tablename__ = "tenant_channel_account"
    __table_args__ = (
        # One default account per (tenant, channel) among live rows.
        Index(
            "uq_channel_account_default", "tenant_id", "channel",
            unique=True,
            postgresql_where=text("is_default = true AND deletion_state = 'ACTIVE'"),
        ),
        Index("ix_channel_account_lookup", "tenant_id", "channel", "is_active"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid7)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("tenant.id"), nullable=False, index=True)
    owner_user_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("app_user.id"), nullable=True)

    channel: Mapped[str] = mapped_column(String(20), nullable=False)        # SMS | TELEGRAM | WHATSAPP
    provider: Mapped[str] = mapped_column(String(40), nullable=False)       # telegram_bot | viva_armenia | whatsapp_cloud
    display_name: Mapped[str] = mapped_column(String(160), nullable=False)
    sender_id: Mapped[str | None] = mapped_column(String(255), nullable=True)   # from-number / bot @username / WABA phone id

    # Fernet-encrypted credentials (the DB never sees plaintext). Which fields a provider uses:
    #   telegram_bot  → secret_token = bot token
    #   viva_armenia  → secret_token = API key/token,  secret_extra = password (if any)
    #   whatsapp_cloud→ secret_token = access token,    config = {phone_number_id, waba_id}
    secret_token: Mapped[str | None] = mapped_column(EncryptedString(), nullable=True)
    secret_extra: Mapped[str | None] = mapped_column(EncryptedString(), nullable=True)
    config: Mapped[dict | None] = mapped_column(JSONB, nullable=True)        # non-secret provider params

    is_default: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, server_default="true")
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="PENDING", server_default="PENDING")
    last_error: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    created_by: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("app_user.id"), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    deletion_state: Mapped[str] = mapped_column(String(20), nullable=False, default="ACTIVE", server_default="ACTIVE")
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
